# 6th_hometask

Домашка по 6му занятию.

Постройте регрессионную модель, используя все те знания, что получили до этого.

Обращаю внимание, что чем подробней вы выполняете задание, тем больше понимания у меня ваших скиллов 🙂

MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

MSZoning: Identifies the general zoning classification of the sale.
		
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM	Residential Medium Density
	
LotFrontage: Linear feet of street connected to property

LotArea: Lot size in square feet

Street: Type of road access to property

       Grvl	Gravel	
       Pave	Paved
       	
Alley: Type of alley access to property

       Grvl	Gravel
       Pave	Paved
       NA 	No alley access
		
LotShape: General shape of property

       Reg	Regular	
       IR1	Slightly irregular
       IR2	Moderately Irregular
       IR3	Irregular
       
LandContour: Flatness of the property

       Lvl	Near Flat/Level	
       Bnk	Banked - Quick and significant rise from street grade to building
       HLS	Hillside - Significant slope from side to side
       Low	Depression
		
Utilities: Type of utilities available
		
       AllPub	All public Utilities (E,G,W,& S)	
       NoSewr	Electricity, Gas, and Water (Septic Tank)
       NoSeWa	Electricity and Gas Only
       ELO	Electricity only	
	
LotConfig: Lot configuration

       Inside	Inside lot
       Corner	Corner lot
       CulDSac	Cul-de-sac
       FR2	Frontage on 2 sides of property
       FR3	Frontage on 3 sides of property
	
LandSlope: Slope of property
		
       Gtl	Gentle slope
       Mod	Moderate Slope	
       Sev	Severe Slope
	
Neighborhood: Physical locations within Ames city limits

       Blmngtn	Bloomington Heights
       Blueste	Bluestem
       BrDale	Briardale
       BrkSide	Brookside
       ClearCr	Clear Creek
       CollgCr	College Creek
       Crawfor	Crawford
       Edwards	Edwards
       Gilbert	Gilbert
       IDOTRR	Iowa DOT and Rail Road
       MeadowV	Meadow Village
       Mitchel	Mitchell
       Names	North Ames
       NoRidge	Northridge
       NPkVill	Northpark Villa
       NridgHt	Northridge Heights
       NWAmes	Northwest Ames
       OldTown	Old Town
       SWISU	South & West of Iowa State University
       Sawyer	Sawyer
       SawyerW	Sawyer West
       Somerst	Somerset
       StoneBr	Stone Brook
       Timber	Timberland
       Veenker	Veenker
			
Condition1: Proximity to various conditions
	
       Artery	Adjacent to arterial street
       Feedr	Adjacent to feeder street	
       Norm	Normal	
       RRNn	Within 200' of North-South Railroad
       RRAn	Adjacent to North-South Railroad
       PosN	Near positive off-site feature--park, greenbelt, etc.
       PosA	Adjacent to postive off-site feature
       RRNe	Within 200' of East-West Railroad
       RRAe	Adjacent to East-West Railroad
	
Condition2: Proximity to various conditions (if more than one is present)
		
       Artery	Adjacent to arterial street
       Feedr	Adjacent to feeder street	
       Norm	Normal	
       RRNn	Within 200' of North-South Railroad
       RRAn	Adjacent to North-South Railroad
       PosN	Near positive off-site feature--park, greenbelt, etc.
       PosA	Adjacent to postive off-site feature
       RRNe	Within 200' of East-West Railroad
       RRAe	Adjacent to East-West Railroad
	
BldgType: Type of dwelling
		
       1Fam	Single-family Detached	
       2FmCon	Two-family Conversion; originally built as one-family dwelling
       Duplx	Duplex
       TwnhsE	Townhouse End Unit
       TwnhsI	Townhouse Inside Unit
	
HouseStyle: Style of dwelling
	
       1Story	One story
       1.5Fin	One and one-half story: 2nd level finished
       1.5Unf	One and one-half story: 2nd level unfinished
       2Story	Two story
       2.5Fin	Two and one-half story: 2nd level finished
       2.5Unf	Two and one-half story: 2nd level unfinished
       SFoyer	Split Foyer
       SLvl	Split Level
	
OverallQual: Rates the overall material and finish of the house

       10	Very Excellent
       9	Excellent
       8	Very Good
       7	Good
       6	Above Average
       5	Average
       4	Below Average
       3	Fair
       2	Poor
       1	Very Poor
	
OverallCond: Rates the overall condition of the house

       10	Very Excellent
       9	Excellent
       8	Very Good
       7	Good
       6	Above Average	
       5	Average
       4	Below Average	
       3	Fair
       2	Poor
       1	Very Poor
		
YearBuilt: Original construction date

YearRemodAdd: Remodel date (same as construction date if no remodeling or additions)

RoofStyle: Type of roof

       Flat	Flat
       Gable	Gable
       Gambrel	Gabrel (Barn)
       Hip	Hip
       Mansard	Mansard
       Shed	Shed
		
RoofMatl: Roof material

       ClyTile	Clay or Tile
       CompShg	Standard (Composite) Shingle
       Membran	Membrane
       Metal	Metal
       Roll	Roll
       Tar&Grv	Gravel & Tar
       WdShake	Wood Shakes
       WdShngl	Wood Shingles
		
Exterior1st: Exterior covering on house

       AsbShng	Asbestos Shingles
       AsphShn	Asphalt Shingles
       BrkComm	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       CemntBd	Cement Board
       HdBoard	Hard Board
       ImStucc	Imitation Stucco
       MetalSd	Metal Siding
       Other	Other
       Plywood	Plywood
       PreCast	PreCast	
       Stone	Stone
       Stucco	Stucco
       VinylSd	Vinyl Siding
       Wd Sdng	Wood Siding
       WdShing	Wood Shingles
	
Exterior2nd: Exterior covering on house (if more than one material)

       AsbShng	Asbestos Shingles
       AsphShn	Asphalt Shingles
       BrkComm	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       CemntBd	Cement Board
       HdBoard	Hard Board
       ImStucc	Imitation Stucco
       MetalSd	Metal Siding
       Other	Other
       Plywood	Plywood
       PreCast	PreCast
       Stone	Stone
       Stucco	Stucco
       VinylSd	Vinyl Siding
       Wd Sdng	Wood Siding
       WdShing	Wood Shingles
	
MasVnrType: Masonry veneer type

       BrkCmn	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       None	None
       Stone	Stone
	
MasVnrArea: Masonry veneer area in square feet

ExterQual: Evaluates the quality of the material on the exterior 
		
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       Po	Poor
		
ExterCond: Evaluates the present condition of the material on the exterior
		
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       Po	Poor
		
Foundation: Type of foundation
		
       BrkTil	Brick & Tile
       CBlock	Cinder Block
       PConc	Poured Contrete	
       Slab	Slab
       Stone	Stone
       Wood	Wood
		
BsmtQual: Evaluates the height of the basement

       Ex	Excellent (100+ inches)	
       Gd	Good (90-99 inches)
       TA	Typical (80-89 inches)
       Fa	Fair (70-79 inches)
       Po	Poor (<70 inches
       NA	No Basement
		
BsmtCond: Evaluates the general condition of the basement

       Ex	Excellent
       Gd	Good
       TA	Typical - slight dampness allowed
       Fa	Fair - dampness or some cracking or settling
       Po	Poor - Severe cracking, settling, or wetness
       NA	No Basement
	
BsmtExposure: Refers to walkout or garden level walls

       Gd	Good Exposure
       Av	Average Exposure (split levels or foyers typically score average or above)	
       Mn	Mimimum Exposure
       No	No Exposure
       NA	No Basement
	
BsmtFinType1: Rating of basement finished area

       GLQ	Good Living Quarters
       ALQ	Average Living Quarters
       BLQ	Below Average Living Quarters	
       Rec	Average Rec Room
       LwQ	Low Quality
       Unf	Unfinshed
       NA	No Basement
		
BsmtFinSF1: Type 1 finished square feet

BsmtFinType2: Rating of basement finished area (if multiple types)

       GLQ	Good Living Quarters
       ALQ	Average Living Quarters
       BLQ	Below Average Living Quarters	
       Rec	Average Rec Room
       LwQ	Low Quality
       Unf	Unfinshed
       NA	No Basement

BsmtFinSF2: Type 2 finished square feet

BsmtUnfSF: Unfinished square feet of basement area

TotalBsmtSF: Total square feet of basement area

Heating: Type of heating
		
       Floor	Floor Furnace
       GasA	Gas forced warm air furnace
       GasW	Gas hot water or steam heat
       Grav	Gravity furnace	
       OthW	Hot water or steam heat other than gas
       Wall	Wall furnace
		
HeatingQC: Heating quality and condition

       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       Po	Poor
		
CentralAir: Central air conditioning

       N	No
       Y	Yes
		
Electrical: Electrical system

       SBrkr	Standard Circuit Breakers & Romex
       FuseA	Fuse Box over 60 AMP and all Romex wiring (Average)	
       FuseF	60 AMP Fuse Box and mostly Romex wiring (Fair)
       FuseP	60 AMP Fuse Box and mostly knob & tube wiring (poor)
       Mix	Mixed
		
1stFlrSF: First Floor square feet
 
2ndFlrSF: Second floor square feet

LowQualFinSF: Low quality finished square feet (all floors)

GrLivArea: Above grade (ground) living area square feet

BsmtFullBath: Basement full bathrooms

BsmtHalfBath: Basement half bathrooms

FullBath: Full bathrooms above grade

HalfBath: Half baths above grade

Bedroom: Bedrooms above grade (does NOT include basement bedrooms)

Kitchen: Kitchens above grade

KitchenQual: Kitchen quality

       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       	
TotRmsAbvGrd: Total rooms above grade (does not include bathrooms)

Functional: Home functionality (Assume typical unless deductions are warranted)

       Typ	Typical Functionality
       Min1	Minor Deductions 1
       Min2	Minor Deductions 2
       Mod	Moderate Deductions
       Maj1	Major Deductions 1
       Maj2	Major Deductions 2
       Sev	Severely Damaged
       Sal	Salvage only
		
Fireplaces: Number of fireplaces

FireplaceQu: Fireplace quality

       Ex	Excellent - Exceptional Masonry Fireplace
       Gd	Good - Masonry Fireplace in main level
       TA	Average - Prefabricated Fireplace in main living area or Masonry Fireplace in basement
       Fa	Fair - Prefabricated Fireplace in basement
       Po	Poor - Ben Franklin Stove
       NA	No Fireplace
		
GarageType: Garage location
		
       2Types	More than one type of garage
       Attchd	Attached to home
       Basment	Basement Garage
       BuiltIn	Built-In (Garage part of house - typically has room above garage)
       CarPort	Car Port
       Detchd	Detached from home
       NA	No Garage
		
GarageYrBlt: Year garage was built
		
GarageFinish: Interior finish of the garage

       Fin	Finished
       RFn	Rough Finished	
       Unf	Unfinished
       NA	No Garage
		
GarageCars: Size of garage in car capacity

GarageArea: Size of garage in square feet

GarageQual: Garage quality

       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       NA	No Garage
		
GarageCond: Garage condition

       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       NA	No Garage
		
PavedDrive: Paved driveway

       Y	Paved 
       P	Partial Pavement
       N	Dirt/Gravel
		
WoodDeckSF: Wood deck area in square feet

OpenPorchSF: Open porch area in square feet

EnclosedPorch: Enclosed porch area in square feet

3SsnPorch: Three season porch area in square feet

ScreenPorch: Screen porch area in square feet

PoolArea: Pool area in square feet

PoolQC: Pool quality
		
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       NA	No Pool
		
Fence: Fence quality
		
       GdPrv	Good Privacy
       MnPrv	Minimum Privacy
       GdWo	Good Wood
       MnWw	Minimum Wood/Wire
       NA	No Fence
	
MiscFeature: Miscellaneous feature not covered in other categories
		
       Elev	Elevator
       Gar2	2nd Garage (if not described in garage section)
       Othr	Other
       Shed	Shed (over 100 SF)
       TenC	Tennis Court
       NA	None
		
MiscVal: $Value of miscellaneous feature

MoSold: Month Sold (MM)

YrSold: Year Sold (YYYY)

SaleType: Type of sale
		
       WD 	Warranty Deed - Conventional
       CWD	Warranty Deed - Cash
       VWD	Warranty Deed - VA Loan
       New	Home just constructed and sold
       COD	Court Officer Deed/Estate
       Con	Contract 15% Down payment regular terms
       ConLw	Contract Low Down payment and low interest
       ConLI	Contract Low Interest
       ConLD	Contract Low Down
       Oth	Other
		
SaleCondition: Condition of sale

       Normal	Normal Sale
       Abnorml	Abnormal Sale -  trade, foreclosure, short sale
       AdjLand	Adjoining Land Purchase
       Alloca	Allocation - two linked properties with separate deeds, typically condo with a garage unit	
       Family	Sale between family members
       Partial	Home was not completed when last assessed (associated with New Homes)


## Import Section

In [1]:
import numpy as np
import pandas as pd

from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

from sklearn.linear_model import LogisticRegression

from sklearn.metrics import classification_report
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error

from matplotlib import pyplot as plt
import seaborn as sn

import catboost

## Path Section

In [2]:
PATH_DATA = r'house_prices.csv'

## Exploratory Data Analysis

In [3]:
df_data = pd.read_csv(PATH_DATA)

In [4]:
df_data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [5]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [6]:
df_data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

### Exploratory Data Analysis - Вывод:

##### Ввиду большого количества признаков, 80 столбцов, и малого количества записей для обучения, 1460 векторов, велик риск, что модель не успеет обучиться при данных соотношениях. Для решения этой проблемы модель будет обучаться на ограниченном списке наиболее важных признаков.

##### Такие признаки как "BsmtQual", "ExterCond"  и другие, имеющие значения: "Excellent", "Good", "Average" и т.д. могут быть обработаны LabelEncoder, так как одзначно известно, что значение признака "Excellent" лучше значения "Good", а значение "Good" лучше значения "Average". Однако при кодировании будут расставлены значения с шагом 1, что может быть ошибочно, так как разные значения могут давать разные разницы влияния на целевую переменную.

##### Также для понимажения размерности данных (снижения количества признаков) такие признаки, как "BsmtQual", описывающий изначальное качество фундамента, и "BsmtCond", описывающий состояние фундамента, могут быть перекодированы в один новый признак с минимальными потерями информатиности (Если не ошибаюсь, что-то вроде метода опорных векторов и понижения размерности).

##### Таким образом, применение OneHotEncoder препочтительнее для сохранения информативности данных. Однако это приведёт к ещё большему количеству признаков, что крайне затруднит обучение модели на небольшом наборе данных

## Data Preprocessing

##### В первую очередь оставим признаки с наибольшим теоретическим потенциалом влияния на целевую переменную.

##### В дальнейшем хотелось бы рассмотреть результативности разных признаков, то есть экспериментальным путём выявить какие комбинации признаков или какие комбинации перекодированных признаков дадут лучшую метрику, но к сожалению не успеваю.

In [7]:
features = ['MSSubClass', 'MSZoning', 'LotArea', 'Utilities', 'BldgType', 'OverallQual', 'OverallCond', 'YearBuilt',
            'ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond', 'HeatingQC', 'KitchenQual', 'SaleType', 'SaleCondition',
            'SalePrice']

X = df_data[features].copy()
X['BsmtQual'] = X['BsmtQual'].fillna('NA').copy()
X['BsmtCond'] = X['BsmtCond'].fillna('NA').copy()

In [8]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   MSSubClass     1460 non-null   int64 
 1   MSZoning       1460 non-null   object
 2   LotArea        1460 non-null   int64 
 3   Utilities      1460 non-null   object
 4   BldgType       1460 non-null   object
 5   OverallQual    1460 non-null   int64 
 6   OverallCond    1460 non-null   int64 
 7   YearBuilt      1460 non-null   int64 
 8   ExterQual      1460 non-null   object
 9   ExterCond      1460 non-null   object
 10  BsmtQual       1460 non-null   object
 11  BsmtCond       1460 non-null   object
 12  HeatingQC      1460 non-null   object
 13  KitchenQual    1460 non-null   object
 14  SaleType       1460 non-null   object
 15  SaleCondition  1460 non-null   object
 16  SalePrice      1460 non-null   int64 
dtypes: int64(6), object(11)
memory usage: 194.0+ KB


## Model Building

##### К сожалению, не успеваю выполнить работу должным образом. Поэтому решил успеть реализовать обработку при помощи CatBoost, так как давно хотел попробовать. Преимущество этого алгоритма, помимо обширного перечня настроек, - он может работать с категориальными признаками без обработки.

In [9]:
%%time

X_train, X_test, y_train, y_test = train_test_split(X.drop(columns='SalePrice'), X['SalePrice'], random_state=1)

dtypes = X_train.dtypes
cat_features = pd.DataFrame(data=dtypes)
cat_features = cat_features[cat_features[0] == 'object'].copy()
cat_features = list(cat_features.index)
cat_features

dtypes = X_train.dtypes
int_features = pd.DataFrame(data=dtypes)
int_features = int_features[int_features[0] != 'object'].copy()
int_features = list(int_features.index)
int_features

X_train[int_features] = X_train[int_features].astype(np.int).copy()
X_test[int_features] = X_test[int_features].astype(np.int).copy()

cb = catboost.CatBoostRegressor(cat_features=cat_features)

cb.fit(X=X_train, y=y_train)
y_pred = cb.predict(X_test)

Learning rate set to 0.041534
0:	learn: 76391.5647046	total: 59.8ms	remaining: 59.8s
1:	learn: 74431.9186751	total: 71.7ms	remaining: 35.8s
2:	learn: 72608.8401054	total: 84.2ms	remaining: 28s
3:	learn: 70954.5988730	total: 93.7ms	remaining: 23.3s
4:	learn: 69248.4510495	total: 106ms	remaining: 21s
5:	learn: 67617.5596056	total: 115ms	remaining: 19.1s
6:	learn: 66137.4977703	total: 126ms	remaining: 17.8s
7:	learn: 64667.3766825	total: 138ms	remaining: 17.1s
8:	learn: 63282.1573416	total: 148ms	remaining: 16.3s


<timed exec>:15: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
<timed exec>:16: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.

9:	learn: 61899.0038950	total: 162ms	remaining: 16s
10:	learn: 60568.3546847	total: 174ms	remaining: 15.7s
11:	learn: 59495.1400958	total: 184ms	remaining: 15.2s
12:	learn: 58405.4766556	total: 196ms	remaining: 14.9s
13:	learn: 57284.9432702	total: 207ms	remaining: 14.6s
14:	learn: 56203.7333047	total: 217ms	remaining: 14.3s
15:	learn: 55266.9665379	total: 226ms	remaining: 13.9s
16:	learn: 54295.2117056	total: 237ms	remaining: 13.7s
17:	learn: 53387.0359004	total: 244ms	remaining: 13.3s
18:	learn: 52525.2816904	total: 252ms	remaining: 13s
19:	learn: 51785.7374269	total: 262ms	remaining: 12.8s
20:	learn: 51025.0723960	total: 272ms	remaining: 12.7s
21:	learn: 50365.2606810	total: 279ms	remaining: 12.4s
22:	learn: 49644.2429898	total: 289ms	remaining: 12.3s
23:	learn: 48983.2692381	total: 300ms	remaining: 12.2s
24:	learn: 48343.1542668	total: 306ms	remaining: 11.9s
25:	learn: 47748.7543903	total: 314ms	remaining: 11.8s
26:	learn: 47170.6325284	total: 323ms	remaining: 11.7s
27:	learn: 4660

160:	learn: 30284.5412134	total: 1.64s	remaining: 8.57s
161:	learn: 30247.1576374	total: 1.65s	remaining: 8.56s
162:	learn: 30239.6021836	total: 1.66s	remaining: 8.54s
163:	learn: 30207.3539364	total: 1.67s	remaining: 8.53s
164:	learn: 30172.4127356	total: 1.68s	remaining: 8.52s
165:	learn: 30172.4085315	total: 1.69s	remaining: 8.47s
166:	learn: 30144.9155029	total: 1.7s	remaining: 8.46s
167:	learn: 30118.3302409	total: 1.71s	remaining: 8.45s
168:	learn: 30072.6462246	total: 1.72s	remaining: 8.44s
169:	learn: 30036.4876144	total: 1.73s	remaining: 8.42s
170:	learn: 29999.5946503	total: 1.74s	remaining: 8.41s
171:	learn: 29980.7526955	total: 1.74s	remaining: 8.4s
172:	learn: 29966.2960214	total: 1.75s	remaining: 8.38s
173:	learn: 29891.2546170	total: 1.76s	remaining: 8.37s
174:	learn: 29871.1213497	total: 1.77s	remaining: 8.37s
175:	learn: 29820.8900654	total: 1.78s	remaining: 8.35s
176:	learn: 29807.8344239	total: 1.79s	remaining: 8.34s
177:	learn: 29751.8221559	total: 1.8s	remaining: 8

311:	learn: 26964.8730944	total: 3.13s	remaining: 6.91s
312:	learn: 26959.3875956	total: 3.14s	remaining: 6.9s
313:	learn: 26939.0699705	total: 3.15s	remaining: 6.89s
314:	learn: 26935.9148054	total: 3.16s	remaining: 6.88s
315:	learn: 26932.9066493	total: 3.17s	remaining: 6.87s
316:	learn: 26930.1349245	total: 3.18s	remaining: 6.86s
317:	learn: 26917.1850740	total: 3.19s	remaining: 6.85s
318:	learn: 26913.5600123	total: 3.2s	remaining: 6.84s
319:	learn: 26867.3287916	total: 3.21s	remaining: 6.83s
320:	learn: 26854.5138669	total: 3.23s	remaining: 6.82s
321:	learn: 26835.6666620	total: 3.23s	remaining: 6.81s
322:	learn: 26807.3672037	total: 3.25s	remaining: 6.8s
323:	learn: 26779.8854096	total: 3.25s	remaining: 6.79s
324:	learn: 26765.6037515	total: 3.27s	remaining: 6.78s
325:	learn: 26762.4898379	total: 3.27s	remaining: 6.77s
326:	learn: 26760.1574083	total: 3.29s	remaining: 6.76s
327:	learn: 26758.9364970	total: 3.29s	remaining: 6.75s
328:	learn: 26739.5505611	total: 3.3s	remaining: 6.

468:	learn: 24658.7574841	total: 4.68s	remaining: 5.3s
469:	learn: 24626.1997666	total: 4.69s	remaining: 5.29s
470:	learn: 24615.8746970	total: 4.7s	remaining: 5.28s
471:	learn: 24598.8749696	total: 4.71s	remaining: 5.27s
472:	learn: 24586.9997623	total: 4.72s	remaining: 5.26s
473:	learn: 24582.4328756	total: 4.73s	remaining: 5.25s
474:	learn: 24571.2141425	total: 4.74s	remaining: 5.24s
475:	learn: 24558.5062243	total: 4.75s	remaining: 5.23s
476:	learn: 24541.1888987	total: 4.76s	remaining: 5.22s
477:	learn: 24539.4426461	total: 4.77s	remaining: 5.21s
478:	learn: 24535.0979389	total: 4.78s	remaining: 5.2s
479:	learn: 24509.6734639	total: 4.79s	remaining: 5.19s
480:	learn: 24493.0163504	total: 4.8s	remaining: 5.18s
481:	learn: 24488.9757767	total: 4.81s	remaining: 5.17s
482:	learn: 24471.1523682	total: 4.82s	remaining: 5.16s
483:	learn: 24458.4641297	total: 4.83s	remaining: 5.14s
484:	learn: 24452.0339707	total: 4.83s	remaining: 5.13s
485:	learn: 24422.1637539	total: 4.84s	remaining: 5.

628:	learn: 22620.2966484	total: 6.27s	remaining: 3.7s
629:	learn: 22595.9473571	total: 6.28s	remaining: 3.69s
630:	learn: 22581.3684329	total: 6.29s	remaining: 3.68s
631:	learn: 22575.0856657	total: 6.3s	remaining: 3.67s
632:	learn: 22572.2922137	total: 6.31s	remaining: 3.66s
633:	learn: 22565.5393442	total: 6.32s	remaining: 3.65s
634:	learn: 22553.8386012	total: 6.33s	remaining: 3.64s
635:	learn: 22538.4742667	total: 6.34s	remaining: 3.63s
636:	learn: 22533.7837573	total: 6.35s	remaining: 3.62s
637:	learn: 22528.5883443	total: 6.36s	remaining: 3.61s
638:	learn: 22523.4456827	total: 6.37s	remaining: 3.6s
639:	learn: 22510.3954060	total: 6.38s	remaining: 3.59s
640:	learn: 22498.5356687	total: 6.39s	remaining: 3.58s
641:	learn: 22498.3113648	total: 6.4s	remaining: 3.57s
642:	learn: 22492.4516422	total: 6.41s	remaining: 3.56s
643:	learn: 22480.3401445	total: 6.42s	remaining: 3.55s
644:	learn: 22464.2607513	total: 6.43s	remaining: 3.54s
645:	learn: 22463.9325530	total: 6.44s	remaining: 3.

785:	learn: 21204.0716535	total: 7.81s	remaining: 2.13s
786:	learn: 21192.7670040	total: 7.82s	remaining: 2.12s
787:	learn: 21183.8871951	total: 7.83s	remaining: 2.11s
788:	learn: 21178.4496800	total: 7.84s	remaining: 2.1s
789:	learn: 21173.0285323	total: 7.85s	remaining: 2.09s
790:	learn: 21160.4120870	total: 7.86s	remaining: 2.08s
791:	learn: 21149.1874020	total: 7.87s	remaining: 2.07s
792:	learn: 21144.7280837	total: 7.88s	remaining: 2.06s
793:	learn: 21141.5645362	total: 7.89s	remaining: 2.05s
794:	learn: 21137.4760226	total: 7.9s	remaining: 2.04s
795:	learn: 21130.9332978	total: 7.91s	remaining: 2.03s
796:	learn: 21125.4357611	total: 7.92s	remaining: 2.02s
797:	learn: 21125.0236883	total: 7.92s	remaining: 2s
798:	learn: 21121.0997967	total: 7.93s	remaining: 2s
799:	learn: 21107.6928315	total: 7.94s	remaining: 1.99s
800:	learn: 21107.0866501	total: 7.95s	remaining: 1.98s
801:	learn: 21092.6306515	total: 7.96s	remaining: 1.97s
802:	learn: 21082.6181198	total: 7.97s	remaining: 1.96s


942:	learn: 20071.4524025	total: 9.36s	remaining: 566ms
943:	learn: 20065.4177294	total: 9.37s	remaining: 556ms
944:	learn: 20057.3117611	total: 9.38s	remaining: 546ms
945:	learn: 20048.5351267	total: 9.39s	remaining: 536ms
946:	learn: 20038.6933936	total: 9.4s	remaining: 526ms
947:	learn: 20028.9375911	total: 9.41s	remaining: 516ms
948:	learn: 20027.0895667	total: 9.42s	remaining: 506ms
949:	learn: 19998.9142954	total: 9.43s	remaining: 496ms
950:	learn: 19994.3209410	total: 9.44s	remaining: 486ms
951:	learn: 19991.6626464	total: 9.45s	remaining: 476ms
952:	learn: 19982.6262803	total: 9.46s	remaining: 466ms
953:	learn: 19968.2229178	total: 9.47s	remaining: 457ms
954:	learn: 19948.6666524	total: 9.48s	remaining: 447ms
955:	learn: 19946.4435190	total: 9.49s	remaining: 437ms
956:	learn: 19933.2601359	total: 9.5s	remaining: 427ms
957:	learn: 19926.7567379	total: 9.51s	remaining: 417ms
958:	learn: 19919.0482597	total: 9.52s	remaining: 407ms
959:	learn: 19910.3645910	total: 9.53s	remaining: 

In [10]:
r2_score(y_test, y_pred)

0.7986926655472831

In [11]:
mean_absolute_error(y_test,y_pred)

23293.65088992715

In [12]:
y_test.describe()

count       365.000000
mean     176725.512329
std       82046.760890
min       40000.000000
25%      125500.000000
50%      160000.000000
75%      205000.000000
max      745000.000000
Name: SalePrice, dtype: float64

##### R2 метрика составила ~0,8, что вроде бы неплохой результат.

##### Показатели MAE говорят о средней ошибке порядка 23000, что довольно существенно для значений в пределах первух двух квартилей.